### PR TITLE
Fix DataGridView.OnRowHeadersWidthSizeModeChanged event invocation

### LINKED ISSUE #3371: DataGridView.OnRowHeadersWidthSizeModeChanged has bad check for RowHeadersWidthSizeModeChanged event
* .NET Core Version: 
master

* Have you experienced this same bug with .NET Framework?: 
Yes

**Problem description:**

In `OnRowHeadersWidthSizeModeChanged` we have the following check

```cs
protected virtual void OnRowHeadersWidthSizeModeChanged(DataGridViewAutoSizeModeEventArgs e)
{
    ...

    if (Events[EVENT_DATAGRIDVIEWROWHEADERSWIDTHSIZEMODECHANGED] is EventHandler eh && !dataGridViewOper[DATAGRIDVIEWOPER_inDispose] && !IsDisposed)
    {
        eh(this, e);
    }
}
```

Note that we write `Events[EVENT_DATAGRIDVIEWROWHEADERSWIDTHSIZEMODECHANGED] is EventHandler eh`

however, `RowHeadersWidthSizeModeChanged` is not an `EventHandler`, but `DataGridViewAutoSizeModeEventHandler`
```cs
[SRCategory(nameof(SR.CatPropertyChanged))]
[SRDescription(nameof(SR.DataGridView_RowHeadersWidthSizeModeChangedDescr))]
public event DataGridViewAutoSizeModeEventHandler RowHeadersWidthSizeModeChanged
{
    add => Events.AddHandler(EVENT_DATAGRIDVIEWROWHEADERSWIDTHSIZEMODECHANGED, value);
    remove => Events.RemoveHandler(EVENT_DATAGRIDVIEWROWHEADERSWIDTHSIZEMODECHANGED, value);
}
```

**Expected behavior:**
The `RowHeadersWidthSizeModeChanged` is called. But it is not

**Minimal repro:**
```cs
[WinFormsTheory]
[InlineData(DataGridViewRowHeadersWidthSizeMode.DisableResizing, false)]
[InlineData(DataGridViewRowHeadersWidthSizeMode.AutoSizeToAllHeaders, true)]
[InlineData(DataGridViewRowHeadersWidthSizeMode.AutoSizeToDisplayedHeaders, true)]
[InlineData(DataGridViewRowHeadersWidthSizeMode.AutoSizeToFirstHeader, true)]
public void DataGridView_RowHeadersWidthSizeMode_SetWithHandler_CallsRowHeadersWidthSizeModeChanged(DataGridViewRowHeadersWidthSizeMode value, object expectedPreviousModeAutoSized)
{
    using var control = new DataGridView();
    int callCount = 0;
    DataGridViewAutoSizeModeEventHandler handler = (sender, e) =>
    {
        Assert.Same(control, sender);
        Assert.Equal(expectedPreviousModeAutoSized, e.PreviousModeAutoSized);
        callCount++;
    };
    control.RowHeadersWidthSizeModeChanged += handler;

    // Set different.
    control.RowHeadersWidthSizeMode = value;
    Assert.Equal(value, control.RowHeadersWidthSizeMode);
    Assert.Equal(1, callCount);

    // Set same.
    control.RowHeadersWidthSizeMode = value;
    Assert.Equal(value, control.RowHeadersWidthSizeMode);
    Assert.Equal(1, callCount);

    // Set different.
    expectedPreviousModeAutoSized = false;
    control.RowHeadersWidthSizeMode = DataGridViewRowHeadersWidthSizeMode.EnableResizing;
    Assert.Equal(DataGridViewRowHeadersWidthSizeMode.EnableResizing, control.RowHeadersWidthSizeMode);
    Assert.Equal(2, callCount);

    // Remove handler.
    control.RowHeadersWidthSizeModeChanged -= handler;
    control.RowHeadersWidthSizeMode = value;
    Assert.Equal(value, control.RowHeadersWidthSizeMode);
    Assert.Equal(2, callCount);
}
```

We fail with `callCount` of `0`, instead of `1`

### LINKED ISSUE #3372: DataGridView.OnColumnHeadersHeightChanged forces creation of the handle
* .NET Core Version: 
Master

* Have you experienced this same bug with .NET Framework?: 
Yes

**Problem description:**

Compare `OnColumnHeadersHeightChanged` with 
```cs
protected virtual void OnColumnHeadersHeightChanged(EventArgs e)
{
    if (editingControl != null)
    {
        PositionEditingControl(true, false, false);
    }

    UpdateMouseEnteredCell(null /*HitTestInfo*/, null /*MouseEventArgs*/);

    OnColumnHeadersGlobalAutoSize();

    if (Events[EVENT_DATAGRIDVIEWCOLUMNHEADERSHEIGHTCHANGED] is EventHandler eh && !dataGridViewOper[DATAGRIDVIEWOPER_inDispose] && !IsDisposed)
    {
        eh(this, e);
    }
}
```

with `OnRowHeadersWidthChanged`
```cs
protected virtual void OnRowHeadersWidthChanged(EventArgs e)
{
    if (RowHeadersVisible)
    {
        if (editingControl != null)
        {
            PositionEditingControl(true, false, false);
        }

        if (IsHandleCreated)
        {
            UpdateMouseEnteredCell(null /*HitTestInfo*/, null /*MouseEventArgs*/);
        }

        OnRowHeadersGlobalAutoSize(false /*expandingRows*/);
    }

    if (Events[EVENT_DATAGRIDVIEWROWHEADERSWIDTHCHANGED] is EventHandler eh && !dataGridViewOper[DATAGRIDVIEWOPER_inDispose] && !IsDisposed)
    {
        eh(this, e);
    }
}
```

Notice that `OnRowHeadersWidthChanged` doesn't force the handle creation, as it guards `UpdateMouseEnteredCell` with `IsHandleCreated`.

However, `OnColumnHeadersHeightChanged` doesn't.

Not only does this force handle creation when `ColumnHeadersHeight` is set to something custom, but also causes `NREs` and `ObjectDisposedException`, for example
```
  X System.Windows.Forms.Tests.DataGridViewTests.DataGridView_ColumnHeadersHeight_SetWithHandlerInColumnDisposing_DoesNotCallColumnHeadersHeightChanged(columnHeadersWidthSizeMode: DisableResizing) [1ms]
  Error Message:
   System.ObjectDisposedException : Cannot access a disposed object.
Object name: 'DataGridView'.
  Stack Trace:
     at System.Windows.Forms.Control.CreateHandle()
   at System.Windows.Forms.Control.get_Handle()
   at System.Windows.Forms.Control.PointToClientInternal(Point p)
   at System.Windows.Forms.DataGridView.UpdateMouseEnteredCell(HitTestInfo hti, MouseEventArgs e)
   at System.Windows.Forms.DataGridView.OnColumnHeadersHeightChanged(EventArgs e)
   at System.Windows.Forms.DataGridView.SetColumnHeadersHeightInternal(Int32 columnHeadersHeight, Boolean invalidInAdjustFillingColumns)
   at System.Windows.Forms.DataGridView.set_ColumnHeadersHeight(Int32 value)
   at System.Windows.Forms.Tests.DataGridViewTests.<>c__DisplayClass12_0.<DataGridView_ColumnHeadersHeight_SetWithHandlerInColumnDisposing_DoesNotCallColumnHeadersHeightChanged>b__0(Object sender, EventArgs e) in C:\Users\hughbe\source\repos\ConsoleApp\DataGridViewTests.cs:line 447
   at System.Windows.Forms.DataGridViewColumn.Dispose(Boolean disposing)
   at System.Windows.Forms.DataGridViewBand.Dispose()
   at System.Windows.Forms.Tests.DataGridViewTests.DataGridView_ColumnHeadersHeight_SetWithHandlerInColumnDisposing_DoesNotCallColumnHeadersHeightChanged(DataGridViewColumnHeadersHeightSizeMode columnHeadersWidthSizeMode) in C:\Users\hughbe\source\repos\ConsoleApp\DataGridViewTests.cs:line 461
   
     at System.Windows.Forms.Tests.DataGridViewTests.DataGridView_ColumnHeadersHeight_SetWithHandlerDisposed_DoesNotCallColumnHeadersHeightChanged(DataGridViewColumnHeadersHeightSizeMode columnHeadersWidthSizeMode) in C:\Users\hughbe\source\repos\ConsoleApp\DataGridViewTests.cs:line 342
The active test run was aborted. Reason: Test host process crashed : Unhandled Exception: System.NullReferenceException: Object reference not set to an instance of an object.
   at System.Windows.Forms.DataGridView.LayoutScrollBars()
   at System.Windows.Forms.DataGridView.ComputeLayout()
   at System.Windows.Forms.DataGridView.PerformLayoutPrivate(Boolean useRowShortcut, Boolean computeVisibleRows, Boolean invalidInAdjustFillingColumns, Boolean repositionEditingControl)
   at System.Windows.Forms.DataGridView.OnHandleCreated(EventArgs e)
   at System.Windows.Forms.Control.WmCreate(Message& m)
   at System.Windows.Forms.Control.WndProc(Message& m)
   at System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, Int32 msg, IntPtr wparam, IntPtr lparam)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.ThreadExceptionFixture.OnThreadException(Object sender, ThreadExceptionEventArgs e) in C:\Users\hughbe\source\repos\ConsoleApp\Support\ThreadExceptionFixture.cs:line 25
```

**Expected behavior:**
We should guard this with call `IsHandleCreated`

**Minimal repro:**
Notice we need `expectedIsHandleCreated` in the following test to make it pass
```cs
public static IEnumerable<object[]> ColumnHeadersHeight_Set_TestData()
{
    foreach (bool columnHeadersVisible in new bool[] { true, false })
    {
        foreach (bool autoSize in new bool[] { true, false })
        {
            yield return new object[] { DataGridViewColumnHeadersHeightSizeMode.AutoSize, columnHeadersVisible, autoSize, 4, DefaultColumnHeadersHeight, false };
            yield return new object[] { DataGridViewColumnHeadersHeightSizeMode.AutoSize, columnHeadersVisible, autoSize, DefaultColumnHeadersHeight, DefaultColumnHeadersHeight, false };
            yield return new object[] { DataGridViewColumnHeadersHeightSizeMode.AutoSize, columnHeadersVisible, autoSize, 32768, DefaultColumnHeadersHeight, false };

            foreach (DataGridViewColumnHeadersHeightSizeMode columnHeadersWidthSizeMode in new DataGridViewColumnHeadersHeightSizeMode[] { DataGridViewColumnHeadersHeightSizeMode.EnableResizing, DataGridViewColumnHeadersHeightSizeMode.DisableResizing })
            {
                yield return new object[] { columnHeadersWidthSizeMode, columnHeadersVisible, autoSize, 4, 4, true };
                yield return new object[] { columnHeadersWidthSizeMode, columnHeadersVisible, autoSize, DefaultColumnHeadersHeight, DefaultColumnHeadersHeight, false };
                yield return new object[] { columnHeadersWidthSizeMode, columnHeadersVisible, autoSize, 32768, 32768, true };
            }
        }
    }
}

[WinFormsTheory]
[MemberData(nameof(ColumnHeadersHeight_Set_TestData))]
public void DataGridView_ColumnHeadersHeight_Set_GetReturnsExpected(DataGridViewColumnHeadersHeightSizeMode columnHeadersWidthSizeMode, bool columnHeadersVisible, bool autoSize, int value, int expectedValue, bool expectedIsHandleCreated)
{
    using var control = new DataGridView
    {
        ColumnHeadersHeightSizeMode = columnHeadersWidthSizeMode,
        ColumnHeadersVisible = columnHeadersVisible,
        AutoSize = autoSize
    };
    int layoutCallCount = 0;
    control.Layout += (sender, e) => layoutCallCount++;

    control.ColumnHeadersHeight = value;
    Assert.Equal(expectedValue, control.ColumnHeadersHeight);
    Assert.Equal(0, layoutCallCount);
    Assert.Equal(expectedIsHandleCreated, control.IsHandleCreated);

    // Set same.
    control.ColumnHeadersHeight = value;
    Assert.Equal(expectedValue, control.ColumnHeadersHeight);
    Assert.Equal(0, layoutCallCount);
    Assert.Equal(expectedIsHandleCreated, control.IsHandleCreated);
}
```

### PR DESCRIPTION
**Pull request by hughbe**
## Proposed Changes
- Fix DataGridView.OnRowHeadersWidthSizeModeChanged event invocation
- Fix DataGridView.OnColumnHeaderHeightChanged forcing handle creation

Fixes #3371
Fixes #3372

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/3373)

### CODE REVIEW

File: src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.Methods.cs [hunk 1]
```diff
@@ -14049,7 +14049,10 @@ protected virtual void OnColumnHeadersHeightChanged(EventArgs e)
                 PositionEditingControl(true, false, false);
             }
 
-            UpdateMouseEnteredCell(null /*HitTestInfo*/, null /*MouseEventArgs*/);
+            if (IsHandleCreated)
+            {
+                UpdateMouseEnteredCell(null /*HitTestInfo*/, null /*MouseEventArgs*/);
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.Methods.cs:14054]: There are 9 other call sites for `UpdateMouseEnteredCell` in this file. I presume those need to be conditioned with `if (IsHandleCreated)` too, needn't they?
May be the check should be moved into `UpdateMouseEnteredCell` method itself, what do you think?

Comment (hughbe) [reply src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.Methods.cs:14054]: Yeah that definitely makes more sense, fixed

Comment (hughbe) [reply src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.Methods.cs:14054]: This does have some side-effects like meaning that `DataGridViewHeaderCell.GetSize` doesn't force handle creation, but this is a good thing!

Comment (hughbe) [reply src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.Methods.cs:14054]: Had to rebase to get those tests in

**Pull request merged**